ramips: add support for Netgear R6020

majik-sheff opened the pull request:
This adds support for the Netgear R6020, aka Netgear AC750.

The R6020 appears to be the same hardware as the Netgear R6080,
aka Netgear AC1000, but it has a slightly different flash layout.

Specification:

SoC: MediaTek MT7628 (580 MHz)
Flash: 8 MiB
RAM: 64 MiB
Wireless: 2.4Ghz (builtin) and 5Ghz (MT7612E)
LAN speed: 10/100
LAN ports: 4
WAN speed: 10/100
WAN ports: 1
UART (57600 8N1) on PCB

Installation:

Flashing OpenWRT from stock firmware requires nmrpflash. Use an ethernet
cable to connect to LAN port 1 of the R6020, and power the R6020 off.
From the connected workstation, run
`nmrpflash -i eth0 -f openwrt-ramips-mt76x8-netgear_r6020-squashfs-factory.img`,
replacing eth0 with the appropriate interface (can be identified by
running `nmrpflash -L`). Then power on the R6020. After flashing has finished,
power cycle the R6020, and it will boot into OpenWRT. Once OpenWRT has been
installed, subsequent flashes can use the web interface and sysupgrade files.

Signed-off-by: Tim Thorpe <timfthorpe@gmail.com>

majik-sheff committed:
ramips: add support for Netgear R6020
This adds support for the Netgear R6020, aka Netgear AC750.

The R6020 appears to be the same hardware as the Netgear R6080,
aka Netgear AC1000, but it has a slightly different flash layout.

Specification:

SoC: MediaTek MT7628 (580 MHz)
Flash: 8 MiB
RAM: 64 MiB
Wireless: 2.4Ghz (builtin) and 5Ghz (MT7612E)
LAN speed: 10/100
LAN ports: 4
WAN speed: 10/100
WAN ports: 1
UART (57600 8N1) on PCB

MAC addresses based on vendor firmware:
LAN      *:88   0x4
WAN      *:89
WLAN2    *:88   0x4
WLAN5    *:8a   0x8004

Installation:

Flashing OpenWRT from stock firmware requires nmrpflash. Use an ethernet
cable to connect to LAN port 1 of the R6020, and power the R6020 off.
From the connected workstation, run
`nmrpflash -i eth0 -f openwrt-ramips-mt76x8-netgear_r6020-squashfs-factory.img`,
replacing eth0 with the appropriate interface (can be identified by
running `nmrpflash -L`). Then power on the R6020. After flashing has finished,
power cycle the R6020, and it will boot into OpenWRT. Once OpenWRT has been
installed, subsequent flashes can use the web interface and sysupgrade files.

Signed-off-by: Tim Thorpe <timfthorpe@gmail.com>

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6020.dts:
```diff
@@ -0,0 +1,175 @@
+	gpio {
+		groups = "p0led_an", "p1led_an", "p2led_an",
+			       "p3led_an", "p4led_an", "wdt",
+			       "wled_an", "uart1";
```
uart1 seems unneeded here.

majik-sheff replied:
It was copied from another dts.  I'll remove it if it's unnecessary.

adrianschmutzler replied:
Probably from r6120 which has USB?

majik-sheff replied:
Guilty.  At least for this portion.  It's a bit of a Frankenstein as you can tell.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6020.dts:
```diff
@@ -0,0 +1,175 @@
+&wmac {
+	status = "okay";
+	mtd-mac-address = <&nvram 0x100b0>;
+	mediatek,mtd-eeprom = <&factory 0x20000>;
```
0x20000 is the size of this partition, so this doesn't make sense.

majik-sheff replied:
Fair point.  I'll try it without these entries and see what happens.

adrianschmutzler replied:
Well, without _both_ entries you will use the default mtd-eeprom from mt7628an.dtsi ...

majik-sheff replied:
Removing both entries results in my build of OpenWRT yielding point-for-point match of MACs to the OEM firmware.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6020.dts:
```diff
@@ -0,0 +1,175 @@
+
+&wmac {
+	status = "okay";
+	mtd-mac-address = <&nvram 0x100b0>;
```
Sounds strange that wmac would require mtd-mac-address, while the 5g WiFi gets it from calibration data ...
But maybe it's just your wrong mtd-eeprom, so please test without mtd-mac-address here again.

majik-sheff replied:
Fair point.  I'll try it without these entries and see what happens.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6020.dts:
```diff
@@ -0,0 +1,175 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x8004>;
```
Nothing in 0x28, 0x2e, 0xe000, 0xe006 or somewhere else in factory partition?

majik-sheff replied:
It's possible it was corrupted when I got to it but this is the address where I found the MAC that's on the sticker.

adrianschmutzler replied:
Yes, because that's the location of the 5g MAC address in calibration data. So, now the questions remain 1. whether this MAC address is supposed to be used for LAN as well and 2. whether the appropriate address is also found in one of the locations named above that are designated for the ethernet address on mt76xx platforms.

majik-sheff replied:
The factory section was definitely corrupted.  The address used by the R6080, R6120, etc is the correct location of the LAN MAC.

adrianschmutzler replied:
You won't convince me by just stating that you are right. You will convince me by reporting the content of 0x4, 0x8004, 0x28, 0x2e, 0xe000, 0xe006 of the factory partition (first and last byte of the address are enough, the rest can be removed if desired) and providing the MAC assignment on vendor firmware.

majik-sheff replied:
`root@OpenWrt:~# hexdump -C /dev/mtd1
00000000  28 76 00 02 00 00 00 01  05 88 00 00 00 00 00 00  |(v..............|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000020  00 00 00 00 20 00 00 00  00 0c 43 e1 76 29 00 0c  |.... .....C.v)..|
00000030  43 e1 76 2a 22 34 00 20  ff ff 00 01 00 00 00 00  |C.v*"4. ........|
00000040  00 00 22 00 00 00 00 00  30 00 00 00 00 00 00 00  |..".....0.......|
00000050  84 00 00 94 40 bb c0 ca  26 81 81 81 40 ca 26 80  |....@...&...@.&.|
00000060  80 80 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000000a0  c1 c1 c1 c1 c1 c1 c0 c1  c1 c1 c1 c1 c1 c0 00 00  |................|
000000b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  00 00 00 00 00 00 00 00  00 1a 22 2a 31 35 01 35  |.........."*15.5|
000000d0  39 40 46 4d 7f 7f 7f 00  00 00 00 00 00 00 00 00  |9@FM............|
000000e0  11 1d 11 1d 1c 35 1c 35  1e 35 1e 35 17 19 17 19  |.....5.5.5.5....|
000000f0  02 00 00 00 b2 00 9c 88  0a 00 00 00 00 00 00 00  |................|
00000100  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000120  00 00 00 00 00 00 00 00  00 00 00 00 00 00 77 00  |..............w.|
00000130  11 1d 11 1d 15 7f 15 7f  17 7f 17 7f 10 3b 10 3b  |.............;.;|
00000140  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000200  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00001000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00008000  62 76 01 00 00 00 00 01  05 8a 12 76 c3 14 00 00  |bv.........v....|
00008010  00 00 12 76 c3 14 00 00  00 00 ff ff ff ff ff ff  |...v............|
00008020  ff ff 37 d8 9d 00 ff 7f  fd 9b ff ff ff ff ff ff  |..7.............|
00008030  ff ff ff ff 22 ff 00 20  ff ff b1 01 00 00 00 00  |....".. ........|
00008040  00 00 ff d7 00 00 00 00  00 00 00 00 00 00 e0 01  |................|
00008050  00 00 00 00 00 ae 7f 05  1a 00 00 00 7e 16 1a 00  |............~...|
00008060  00 00 82 e9 1a 00 00 82  e9 22 83 83 81 ed 22 83  |........."....".|
00008070  83 80 f0 22 83 83 81 f6  22 81 81 80 fe 22 81 81  |..."...."...."..|
00008080  82 e2 1a 00 00 82 e2 22  83 83 82 e8 22 83 83 82  |......."...."...|
00008090  e6 22 83 83 83 eb 22 83  83 83 f2 22 83 83 00 54  |."...."...."...T|
000080a0  c3 c3 c4 c4 00 00 c3 c3  c1 82 c3 c3 c1 82 00 00  |................|
000080b0  00 00 c3 c3 c3 00 00 00  00 00 00 00 00 00 85 85  |................|
000080c0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
000080f0  ff ff ff ff ff ff a5 22  22 00 00 00 00 00 00 00  |......."".......|
00008100  0f ca 74 c5 e8 07 30 3d  01 b0 08 26 00 0e 04 15  |..t...0=...&....|
00008110  00 8a 00 40 00 00 00 08  00 9d 08 00 12 c0 00 00  |...@............|
00008120  08 20 04 2a 90 00 00 24  01 04 54 08 d0 a0 28 20  |. .*...$..T...( |
00008130  ff ff ff ff ff ff ff 08  ff ff ff ff ff ff 00 00  |................|
00008140  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
000081e0  c0 81 82 c3 04 45 46 07  08 09 ff ff ff ff ff ff  |.....EF.........|
000081f0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00020000
`
FW gives MACs:
LAN 00:00:00:01:05:88
WAN 00:00:00:01:05:89
WLAN2 00:00:00:01:05:88
WLAN5 00:00:00:01:05:8a

majik-sheff replied:
I thought telling it to use code formatting would respect line breaks.  If that's too much of a mess I'll attach a txt file.

majik-sheff replied:
I made the changes I proposed and checked OpenWRT's macs vs stock firmware and they match.  I've pushed the changes for your review.

adrianschmutzler replied:
I fixed the formatting of your comment. Single backticks can only be used inline, not across lines. If you want to add a code block, you need triple backticks before and after the code.

Thanks for the data, this seems sane and looks correct in your implementation (except for the label MAC address issue). I'm just a little startled by the triple zero bytes. Please add your findings about MAC addresses on vendor FW and locations to the commit message, e.g. in a table like this:

```
MAC addresses based on vendor firmware:
LAN      *:88   0x4
WAN      *:89
WLAN2    *:88   0x4
WLAN5    *:8a   0x8004   (label)
```

majik-sheff replied:
Unfortunately, the original contents of the factory partition appear to have been damaged (probably in one of my early flash attempts) and so the sticker mac is no longer present in memory.  When some of the data was still intact I found the sticker MAC + 4 in the 0x8004 address.  This is why I had the original strangeness for deriving the MAC, because I had apparently already damaged the data sitting at 0x4 before I got to this stage.  Unless someone else has an R6020 they can dump MTD1 out of, I'll have to procure another victim to confirm what the original values were.

As it stands, my configuration produces results that match what the OEM firmware outputs without resorting to any strange math or other tricks.  I removed the redundant alias line you pointed out and squashed again.

adrianschmutzler replied:
Okay, I see. So even with possibly broken content, OEM firmware still told us the correct locations; and this is in line with the other devices as well.
I also think just removing label MAC address in this case is the best option, it can be easily added afterwards at any time (and it is only present for new devices anyway).

majik-sheff replied:
Exactly what I was trying to convey.  Thank you.

adrianschmutzler commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -189,6 +189,25 @@ define Device/mercury_mac1200r-v2
+  IMAGE_SIZE := 7104k
+  DEVICE_VENDOR := NETGEAR
+  DEVICE_MODEL := R6020
+  DEVICE_PACKAGES := kmod-mt76x2 kmod-usb2 kmod-usb-ohci
```
Please decide whether you have USB or not. Currently, you disable it in DTS.

majik-sheff replied:
This was copied from another similar device.  I'll remove it and try again.

adrianschmutzler replied:
So, do you have USB or not?

majik-sheff replied:
No.  This device does not have USB ports.

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -181,6 +182,9 @@ ramips_setup_macs()
 		lan_mac=$(mtd_get_mac_binary factory 0x2e)
 		;;
+	netgear,r6020)
+		wan_mac=$(macaddr_add "$(mtd_get_mac_binary factory 0x8004)" -4)
```
"-4" seems rather unprobable to me. Can you please report the MAC address assignment on vendor firmware? I'd need the last byte for LAN, WAN, 2.4g and 5g interfaces ...

majik-sheff replied:
I'm wondering if this was meant to be the mac address for a wireless device?  I think I might have a partially corrupted config partition.

adrianschmutzler replied:
Well, the art of copying together code from other similar devices is to find out what to keep, what to remove, and what to adjust properly.
The MAC address assignment is one of the latter kind. And to adjust them properly, you first have to find out what you have to adjust them to. So, please, flash back to vendor firmware, check the MAC addresses there as requested above, and then report them here. From this point, we may move forward towards how to implement them properly.

Further reading:
https://openwrt.org/docs/guide-developer/mac.address

majik-sheff replied:
Did that.  See my other comment.

majik-sheff commented:
I have made the needed changes and squashed the commits.  My test build now presents the same MACs for all network interfaces as the factory firmware and without resorting to strange addresses.   I also removed the redundant/wrong leftovers from the donor files and cleaned up where needed.  Please let me know what else is needed.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6020.dts:
```diff
@@ -0,0 +1,173 @@
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+		label-mac-device = &ethernet;
```
Earlier, you stated the label shows the 5g MAC address. In that case, this won't work, but you would have to set label_mac in 02_network. Just look for examples with other devices, e.g. cudy,wr1000 ...

majik-sheff replied:
Ah. I didn't see that.

adrianschmutzler commented:
Thanks, pulled into my staging tree.
